In [1]:
pip install formulaic

Note: you may need to restart the kernel to use updated packages.



[notice] A new release of pip is available: 24.0 -> 24.3.1
[notice] To update, run: C:\Users\Estefany\AppData\Local\Microsoft\WindowsApps\PythonSoftwareFoundation.Python.3.11_qbz5n2kfra8p0\python.exe -m pip install --upgrade pip


In [1]:
import numpy as np
import pandas as pd
from formulaic import Formula
from sklearn.model_selection import train_test_split

# Data analysis

In [2]:
data = pd.read_csv('../data/raw/wage2015_subsample_inference.csv') # Import the data set
list(data.columns)

['wage',
 'lwage',
 'sex',
 'shs',
 'hsg',
 'scl',
 'clg',
 'ad',
 'mw',
 'so',
 'we',
 'ne',
 'exp1',
 'exp2',
 'exp3',
 'exp4',
 'occ',
 'occ2',
 'ind',
 'ind2']

### Missing values:

In [3]:
data.isna().sum()

wage     0
lwage    0
sex      0
shs      0
hsg      0
scl      0
clg      0
ad       0
mw       0
so       0
we       0
ne       0
exp1     0
exp2     0
exp3     0
exp4     0
occ      0
occ2     0
ind      0
ind2     0
dtype: int64

### Descriptive statistics of the variables:

In [4]:
data.describe()

wage        lwage          sex          shs          hsg  \
count  5150.000000  5150.000000  5150.000000  5150.000000  5150.000000   
mean     23.410410     2.970787     0.444466     0.023301     0.243883   
std      21.003016     0.570385     0.496955     0.150872     0.429465   
min       3.021978     1.105912     0.000000     0.000000     0.000000   
25%      13.461538     2.599837     0.000000     0.000000     0.000000   
50%      19.230769     2.956512     0.000000     0.000000     0.000000   
75%      27.777778     3.324236     1.000000     0.000000     0.000000   
max     528.845673     6.270697     1.000000     1.000000     1.000000   

               scl          clg           ad           mw           so  \
count  5150.000000  5150.000000  5150.000000  5150.000000  5150.000000   
mean      0.278058     0.317670     0.137087     0.259612     0.296505   
std       0.448086     0.465616     0.343973     0.438464     0.456761   
min       0.000000     0.000000     0.000000     0.000000     0.000000   
25%       0.000000     0.000000     0.000000     0.000000     0.000000   
50%       0.000000     0.000000     0.000000     0.000000     0.000000   
75%       1.000000     1.000000     0.000000     1.000000     1.000000   
max       1.000000     1.000000     1.000000     1.000000     1.000000   

                we           ne         exp1         exp2         exp3  \
count  5150.000000  5150.000000  5150.000000  5150.000000  5150.000000   
mean      0.216117     0.227767    13.760583     3.018925     8.235867   
std       0.411635     0.419432    10.609465     4.000904    14.488962   
min       0.000000     0.000000     0.000000     0.000000     0.000000   
25%       0.000000     0.000000     5.000000     0.250000     0.125000   
50%       0.000000     0.000000    10.000000     1.000000     1.000000   
75%       0.000000     0.000000    21.000000     4.410000     9.261000   
max       1.000000     1.000000    47.000000    22.090000   103.823000   

              exp4            occ         occ2            ind         ind2  
count  5150.000000    5150.000000  5150.000000    5150.000000  5150.000000  
mean     25.118038    5310.737476    11.670874    6629.154951    13.316893  
std      53.530225   11874.356080     6.966684    5333.443992     5.701019  
min       0.000000      10.000000     1.000000     370.000000     2.000000  
25%       0.062500    1740.000000     5.000000    4880.000000     9.000000  
50%       1.000000    4040.000000    13.000000    7370.000000    14.000000  
75%      19.448100    5610.000000    17.000000    8190.000000    18.000000  
max     487.968100  100000.000000    22.000000  100000.000000    22.000000

In [5]:
percentile = np.arange(0,1,0.05) # more percentiles included
data.describe(percentiles=percentile)

wage        lwage          sex          shs          hsg  \
count  5150.000000  5150.000000  5150.000000  5150.000000  5150.000000   
mean     23.410410     2.970787     0.444466     0.023301     0.243883   
std      21.003016     0.570385     0.496955     0.150872     0.429465   
min       3.021978     1.105912     0.000000     0.000000     0.000000   
0%        3.021978     1.105912     0.000000     0.000000     0.000000   
5%        8.173077     2.100845     0.000000     0.000000     0.000000   
10%       9.615385     2.263364     0.000000     0.000000     0.000000   
15%      11.207532     2.416586     0.000000     0.000000     0.000000   
20%      12.019231     2.486508     0.000000     0.000000     0.000000   
25%      13.461538     2.599837     0.000000     0.000000     0.000000   
30%      14.423077     2.668829     0.000000     0.000000     0.000000   
35%      15.384615     2.733368     0.000000     0.000000     0.000000   
40%      16.826923     2.822980     0.000000     0.000000     0.000000   
45%      17.713942     2.874352     0.000000     0.000000     0.000000   
50%      19.230769     2.956512     0.000000     0.000000     0.000000   
55%      20.192308     3.005302     0.000000     0.000000     0.000000   
60%      21.634615     3.074295     1.000000     0.000000     0.000000   
65%      24.038462     3.179655     1.000000     0.000000     0.000000   
70%      25.641026     3.244194     1.000000     0.000000     0.000000   
75%      27.777778     3.324236     1.000000     0.000000     0.000000   
80%      30.364372     3.413270     1.000000     0.000000     1.000000   
85%      34.188034     3.531876     1.000000     0.000000     1.000000   
90%      39.615385     3.679218     1.000000     0.000000     1.000000   
95%      50.000000     3.912023     1.000000     0.000000     1.000000   
max     528.845673     6.270697     1.000000     1.000000     1.000000   

               scl          clg           ad           mw           so  \
count  5150.000000  5150.000000  5150.000000  5150.000000  5150.000000   
mean      0.278058     0.317670     0.137087     0.259612     0.296505   
std       0.448086     0.465616     0.343973     0.438464     0.456761   
min       0.000000     0.000000     0.000000     0.000000     0.000000   
0%        0.000000     0.000000     0.000000     0.000000     0.000000   
5%        0.000000     0.000000     0.000000     0.000000     0.000000   
10%       0.000000     0.000000     0.000000     0.000000     0.000000   
15%       0.000000     0.000000     0.000000     0.000000     0.000000   
20%       0.000000     0.000000     0.000000     0.000000     0.000000   
25%       0.000000     0.000000     0.000000     0.000000     0.000000   
30%       0.000000     0.000000     0.000000     0.000000     0.000000   
35%       0.000000     0.000000     0.000000     0.000000     0.000000   
40%       0.000000     0.000000     0.000000     0.000000     0.000000   
45%       0.000000     0.000000     0.000000     0.000000     0.000000   
50%       0.000000     0.000000     0.000000     0.000000     0.000000   
55%       0.000000     0.000000     0.000000     0.000000     0.000000   
60%       0.000000     0.000000     0.000000     0.000000     0.000000   
65%       0.000000     0.000000     0.000000     0.000000     0.000000   
70%       0.000000     1.000000     0.000000     0.000000     0.000000   
75%       1.000000     1.000000     0.000000     1.000000     1.000000   
80%       1.000000     1.000000     0.000000     1.000000     1.000000   
85%       1.000000     1.000000     0.000000     1.000000     1.000000   
90%       1.000000     1.000000     1.000000     1.000000     1.000000   
95%       1.000000     1.000000     1.000000     1.000000     1.000000   
max       1.000000     1.000000     1.000000     1.000000     1.000000   

                we           ne         exp1         exp2         exp3  \
count  5150.000000  5150.000000  5150.000000  5150.000000  5150.000000   
mean      0.2161

As most of the variables are dummy variables, only wage is interprated based in these descriptive statistics. The average wage in the sample is 23.410 dollars per hour. 

The 25% richest earn 27.78 dollars per hour or more, while the 25% poorest earn 13.46 or less. 

The median earning per hour is 19.23 dollars, which is lower than the average wage and shows the asymmetry in the wage distribution of the sample. 

### Women with:
* a college graduate degree (clg) 
* have a wage corresponding to the 25% richest of the sample

In [6]:
data_women_clg_75 = data[(data['sex'] == 1) &
                         ((data['clg'] == 1)|(data['ad'] == 1)) &
                         (data['wage']>data['wage'].describe()['75%'])]
data_women_clg_75

wage     lwage  sex  shs  hsg  scl  clg  ad  mw  so  we  ne  exp1  \
4      28.846154  3.361977    1    0    0    0    1   0   0   0   0   1  22.0   
30     42.307692  3.744969    1    0    0    0    0   1   0   0   0   1   0.0   
39     41.208791  3.718652    1    0    0    0    1   0   0   0   0   1  11.0   
54    100.000000  4.605170    1    0    0    0    0   1   0   0   0   1  27.0   
91     33.653846  3.516127    1    0    0    0    1   0   0   0   0   1   4.0   
...          ...       ...  ...  ...  ...  ...  ...  ..  ..  ..  ..  ..   ...   
5085   48.076923  3.872802    1    0    0    0    0   1   0   0   1   0   0.0   
5093   30.145530  3.406037    1    0    0    0    1   0   0   0   1   0  15.0   
5105   29.914530  3.398344    1    0    0    0    1   0   0   0   1   0  39.0   
5131   48.076923  3.872802    1    0    0    0    1   0   0   0   1   0  25.0   
5140   45.546559  3.818735    1    0    0    0    1   0   0   0   1   0   5.0   

       exp2    exp3      exp4     occ  occ2     ind  ind2  
4      4.84  10.648   23.4256  2015.0     6  9470.0    22  
30     0.00   0.000    0.0000  4700.0    16  4795.0     9  
39     1.21   1.331    1.4641  4700.0    16  5790.0     9  
54     7.29  19.683   53.1441   850.0     2  6970.0    12  
91     0.16   0.064    0.0256   710.0     2  7390.0    14  
...     ...     ...       ...     ...   ...     ...   ...  
5085   0.00   0.000    0.0000  3050.0    10  8090.0    18  
5093   2.25   3.375    5.0625   120.0     1  6870.0    12  
5105  15.21  59.319  231.3441  4700.0    16  5380.0     9  
5131   6.25  15.625   39.0625  3255.0    10  8170.0    18  
5140   0.25   0.125    0.0625  3255.0    10  8190.0    18  

[412 rows x 20 columns]

In [7]:
print(f'{data_women_clg_75.shape[0]} women with a college graduate degree (clg) or above have a wage corresponding to the 25% richest of the sample')

412 women with a college graduate degree (clg) or above have a wage corresponding to the 25% richest of the sample


### Men with:
* a high school degree (hsg) 
* have a wage corresponding to the 25% richest of the sample

In [8]:
data_men_hsg_75 = data[(data['sex'] == 1) &
                       ((data['shs'] == 1)|(data['hsg'] == 1)) &
                       (data['wage']>data['wage'].describe()['75%'])]
data_men_hsg_75

wage     lwage  sex  shs  hsg  scl  clg  ad  mw  so  we  ne  exp1  \
267    33.653846  3.516127    1    0    1    0    0   0   0   0   0   1  33.0   
330    79.326923  4.373578    1    0    1    0    0   0   0   0   0   1  26.0   
722    28.846154  3.361977    1    1    0    0    0   0   0   0   0   1   8.0   
726    30.769231  3.426515    1    0    1    0    0   0   0   0   0   1  21.0   
914    31.730769  3.457287    1    0    1    0    0   0   0   0   0   1  24.0   
917    37.179487  3.615757    1    0    1    0    0   0   0   0   0   1  36.0   
1017   30.364372  3.413270    1    0    1    0    0   0   0   0   0   1  21.0   
1044   60.096154  4.095946    1    0    1    0    0   0   0   0   0   1  19.0   
1074   34.722222  3.547380    1    0    1    0    0   0   0   0   0   1  12.0   
1488   31.250000  3.442019    1    0    1    0    0   0   1   0   0   0  31.0   
1553   33.653846  3.516127    1    0    1    0    0   0   1   0   0   0  20.0   
1633   29.807692  3.394766    1    0    1    0    0   0   1   0   0   0  22.0   
1771   28.846154  3.361977    1    0    1    0    0   0   1   0   0   0  29.0   
1939   31.730769  3.457287    1    0    1    0    0   0   1   0   0   0  18.0   
2073   28.846154  3.361977    1    0    1    0    0   0   1   0   0   0  17.0   
3175   47.008547  3.850329    1    1    0    0    0   0   0   1   0   0  23.0   
3239   46.538462  3.840279    1    0    1    0    0   0   0   1   0   0  26.0   
3321   33.653846  3.516127    1    0    1    0    0   0   0   1   0   0  33.0   
3727   58.173077  4.063423    1    0    1    0    0   0   0   1   0   0  37.0   
3770   42.307692  3.744969    1    0    1    0    0   0   0   1   0   0   5.0   
3895   62.500000  4.135167    1    0    1    0    0   0   0   1   0   0  21.0   
4199   33.653846  3.516127    1    0    1    0    0   0   0   0   1   0  22.0   
4262   43.269231  3.767442    1    0    1    0    0   0   0   0   1   0   8.0   
4270   30.769231  3.426515    1    0    1    0    0   0   0   0   1   0   7.0   
4389   33.333333  3.506558    1    0    1    0    0   0   0   0   1   0  35.0   
4488   64.102564  4.160484    1    1    0    0    0   0   0   0   1   0  15.0   
4595  145.960664  4.983337    1    0    1    0    0   0   0   0   1   0  12.0   
4710   29.326923  3.378506    1    0    1    0    0   0   0   0   1   0   9.0   
4826   72.115385  4.278267    1    0    1    0    0   0   0   0   1   0  12.0   
4916   29.567308  3.386669    1    0    1    0    0   0   0   0   1   0  20.0   
4920   37.500000  3.624341    1    0    1    0    0   0   0   0   1   0  31.0   
4995   32.692308  3.487140    1    0    1    0    0   0   0   0   1   0  10.0   
5038   36.057692  3.585120    1    0    1    0    0   0   0   0   1   0  36.0   

       exp2    exp3      exp4     occ  occ2     ind  ind2  
267   10.89  35.937  118.5921  2710.0     9  6570.0    11  
330    6.76  17.576   45.6976  4710.0    16  4170.0     7  
722    0.64   0.512    0.4096  5240.0    17  6970.0    12  
726    4.41   9.261   19.4481  3420.0    10  8190.0    18  
914    5.76  13.824   33.1776  4830.0    16  7670.0    16  
917   12.96  46.656  167.9616   430.0     1  7870.0    17  
1017   4.41   9.261   19.4481  3600.0    11  8270.0    18  
1044   3.61   6.859   13.0321   410.0     1  9570.0    22  
1074   1.44   1.728    2.0736  3500.0    10  8190.0    18  
1488   9.61  29.791   92.3521  3255.0    10  8190.0    18  
1553   4.00   8.000   16.0000  6200.0    19   770.0     4  
1633   4.84  10.648   23.4256   950.0     2  6970.0    12  
1771   8.41  24.389   70.7281  1050.0     3  7380.0    14  
1939   3.24   5.832   10.4976  4830.0    16  7670.0    16  
2073   2.89   4.913    8.3521   530.0     2  3580.0     6  
3175   5.29  12.167   27.9841  1320.0     4  3580.0     6  
3239   6.76  17.576   45.6976    50.0     1  7380.0    14  
3321  10.89  35.937  118.5921  2710.0     9  6570.0    11  
3727  13.69  50.653  187.4161    50.0     1  5380.0     9  
3770   0.25   0.125    0.0625  4720.0    16  4970.0     9

In [9]:
print(f'{data_men_hsg_75.shape[0]} men with a high school graduate degree (hsg) or below have a wage corresponding to the 25% richest of the sample')

33 men with a high school graduate degree (hsg) or below have a wage corresponding to the 25% richest of the sample


## New Dataframes:

* One containing only the log(wage) 
* One containig every variable of the data set but the "wage" related variables

In [10]:
log_wage = pd.DataFrame(data['lwage'])
log_wage


lwage
0     2.263364
1     3.872802
2     2.403126
3     2.634928
4     3.361977
...        ...
5145  2.692546
5146  3.138833
5147  3.649659
5148  3.495508
5149  2.851151

[5150 rows x 1 columns]

In [11]:
vars = [x for x in data.columns if x != 'wage' and x !='lwage']
features = pd.DataFrame(data[vars])
features

sex  shs  hsg  scl  clg  ad  mw  so  we  ne  exp1  exp2    exp3  \
0       1    0    0    0    1   0   0   0   0   1   7.0  0.49   0.343   
1       0    0    0    0    1   0   0   0   0   1  31.0  9.61  29.791   
2       0    0    1    0    0   0   0   0   0   1  18.0  3.24   5.832   
3       1    0    0    0    0   1   0   0   0   1  25.0  6.25  15.625   
4       1    0    0    0    1   0   0   0   0   1  22.0  4.84  10.648   
...   ...  ...  ...  ...  ...  ..  ..  ..  ..  ..   ...   ...     ...   
5145    0    0    0    0    1   0   0   0   1   0   9.0  0.81   0.729   
5146    1    0    0    1    0   0   0   0   1   0  12.0  1.44   1.728   
5147    0    0    0    0    0   1   0   0   1   0  11.0  1.21   1.331   
5148    0    0    1    0    0   0   0   0   1   0  10.0  1.00   1.000   
5149    0    0    0    0    0   1   0   0   1   0  14.0  1.96   2.744   

         exp4     occ  occ2     ind  ind2  
0      0.2401  3600.0    11  8370.0    18  
1     92.3521  3050.0    10  5070.0     9  
2     10.4976  6260.0    19   770.0     4  
3     39.0625   420.0     1  6990.0    12  
4     23.4256  2015.0     6  9470.0    22  
...       ...     ...   ...     ...   ...  
5145   0.6561  4700.0    16  4970.0     9  
5146   2.0736  4110.0    13  8680.0    20  
5147   1.4641  1550.0     4  3680.0     6  
5148   1.0000  2920.0     9  6570.0    11  
5149   3.8416  1610.0     5  7460.0    14  

[5150 rows x 18 columns]

# Data Wrangling

In [12]:
y = data["lwage"].to_numpy().reshape(-1, 1) #array for "y" variable

## Models:

### The Basic Model

$
\text{sex} + \text{exp1} + \text{hsg} + \text{scl} + \text{clg} + \text{ad} + \text{so} + \text{we} + \text{ne} + \text{dummy(occ2)} + \text{dummy(ind2)}
$

Variables:
- ex, hsg, scl, clg, ad, so, we, ne, exp1 
- dummys: occ2, ind2

In [13]:
basic_predictors = ["sex", "hsg", "scl", "clg", "ad", "so", "we", "ne", "exp1", "occ2", "ind2"] # these are our predictor columns

X_basic = pd.get_dummies(data[basic_predictors], columns = ["occ2", "ind2"], drop_first = True, dtype = int) # get the dummies (0 or 1)
#obs: `drop_first = True` to avoid multicolinearity
X_basic["Intercept"] = 1

### The Flexible Model

$
\text{sex} + \text{exp1} + \text{exp2} + \text{exp3} + \text{exp4} +
\text{hsg} + \text{scl} + \text{clg} + \text{ad} + \text{so} + 
\text{we} + \text{ne} + \text{dummy(occ2)} + \text{dummy(ind2)} +
\left( \text{exp1} + \text{exp2} + \text{exp3} + \text{exp4} \right) \times 
\left( \text{hsg} + \text{scl} + \text{clg} + \text{ad} + \text{so} + 
\text{we} + \text{ne} + \text{dummy(occ2)} + \text{dummy(ind2)} \right)
$

In [14]:
flexible_predictors = basic_predictors + ["exp2", "exp3", "exp4"] # new list
X_flexible = pd.get_dummies(data[flexible_predictors], columns = ["occ2", "ind2"], drop_first = True)

indicator_variables = [variable for variable in X_flexible.columns 
                       if any(base_name in variable for base_name in ["hsg", "scl", "clg", "ad", "so", "we", "ne", "occ2", "ind2"])]

for experience_variable in ["exp1", "exp2", "exp3", "exp4"]:
        for indicator_variable in indicator_variables:
                temp_frame = pd.DataFrame({
                        f"{experience_variable}_{indicator_variable}": X_flexible[experience_variable] * X_flexible[indicator_variable]  
                })
                X_flexible = pd.concat((X_flexible, temp_frame), axis = 1)

X_flexible["Intercept"] = 1 

### The extra-flexible model

In [15]:
X_extra_flexible = X_basic.copy()

X_extra_flexible = pd.concat((X_extra_flexible, data[["exp2", "exp3", "exp4"]]), axis = 1)

extra_flexible_base_features = X_extra_flexible.columns[1:],

In [16]:
base_experience_variables = ["exp1", "exp2", "exp3"]
exp_4 = ["exp4"]
education_variables = ["hsg", "scl", "clg", "ad"]
location_variables = ["so", "we", "ne"]
occupation_variables = [variable for variable in X_extra_flexible.columns if "occ2" in variable]
industry_variables = [variable for variable in X_extra_flexible.columns if "ind2" in variable]
variable_groups = [base_experience_variables, exp_4, education_variables, location_variables, occupation_variables, industry_variables]

In [17]:
for variable_group_index, variable_group in enumerate(variable_groups, 1):
        if variable_group == len(variable_groups):
                break

        other_variable_groups = variable_groups[variable_group_index:]

        for variable in variable_group:
                for other_variables in other_variable_groups:
                        temp = X_extra_flexible[other_variables].multiply(X_extra_flexible[variable], axis = "index")
                        temp = temp.add_prefix(f"{variable}")
                        X_extra_flexible = pd.concat(
                                (X_extra_flexible, temp),
                                axis = 1
                        )

In [18]:
temp = pd.DataFrame({"exp8": X_extra_flexible["exp4"] ** 2})
X_extra_flexible = pd.concat((X_extra_flexible, temp), axis = 1)

# Linear Regressions

Split each of the dataframes (basic, flexible and extra-flexible models) into a training sample (80%) and a test sample (20%).

* Regression Basic Model

In [19]:

X_basic_train, X_basic_test, y_basic_train, y_basic_test = train_test_split(X_basic, y, train_size = .8, shuffle = True)

* Regression Flexible Model

In [20]:

X_flexible_train, X_flexible_test, y_flexible_train, y_flexible_test = train_test_split(X_flexible, y, train_size = .8, shuffle = True)

* Regression Extra Flexible Model

In [21]:

X_extra_flexible_train, X_extra_flexible_test, y_extra_flexible_train, y_extra_flexible_test = train_test_split(X_extra_flexible, y, train_size = .8, shuffle = True)

## Estimation

* Regression Basic Model

In [22]:
from sklearn.linear_model import LinearRegression
basic_model = LinearRegression()
basic_model.fit(X_basic_train, y_basic_train)

LinearRegression()

* Regression Flexible Model

In [23]:
flexible_model = LinearRegression()
flexible_model.fit(X_flexible_train, y_flexible_train)

LinearRegression()

* Regression Extra Flexible Model

In [24]:
extra_flexible_model = LinearRegression()
extra_flexible_model.fit(X_extra_flexible_train, y_extra_flexible_train)

LinearRegression()

## Report Result

Report the MSE and the R^2 for both samples of each model, and the adjusted R^2 for the training sample. 

In [25]:
basic_training_mse = ((y_basic_train - basic_model.predict(X_basic_train)) ** 2).mean()
basic_training_R2 = 1 - basic_training_mse / ((y_basic_train - y_basic_train.mean()) ** 2).mean()
basic_training_adjR2 = 1 - basic_training_mse * y_basic_train.shape[0] / (y_basic_train.shape[0] - basic_model.n_features_in_) / ((y_basic_train - y_basic_train.mean()) ** 2).mean()

print(f"Mean square error of basic model: {basic_training_mse: .4f}")
print(f"R2 of basic model: {basic_training_R2: .4f}")
print(f"Adjusted R2 of basic model: {basic_training_adjR2: .4f}")

Mean square error of basic model:  0.2272
R2 of basic model:  0.3062
Adjusted R2 of basic model:  0.2975


In [26]:
basic_testing_mse = ((y_basic_test - basic_model.predict(X_basic_test)) ** 2).mean()
basic_testing_R2 = 1 - basic_testing_mse / ((y_basic_test - y_basic_test.mean()) ** 2).mean()

print(f"Mean square error of basic model: {basic_testing_mse: .4f}")
print(f"R2 of basic model: {basic_testing_R2: .4f}")

Mean square error of basic model:  0.2155
R2 of basic model:  0.3171


In [27]:
flexible_training_mse = ((y_flexible_train - flexible_model.predict(X_flexible_train)) ** 2).mean()
flexible_training_R2 = 1 - flexible_training_mse / ((y_flexible_train - y_flexible_train.mean()) ** 2).mean()
flexible_training_adjR2 = 1 - flexible_training_mse * y_flexible_train.shape[0] / (y_flexible_train.shape[0] - flexible_model.n_features_in_) / ((y_flexible_train - y_flexible_train.mean()) ** 2).mean()

print(f"Mean square error of flexible model: {flexible_training_mse: .4f}")
print(f"R2 of flexible model: {flexible_training_R2: .4f}")
print(f"Adjusted R2 of flexible model: {flexible_training_adjR2: .4f}")

Mean square error of flexible model:  0.2076
R2 of flexible model:  0.3583
Adjusted R2 of flexible model:  0.3176


In [28]:
flexible_testing_mse = ((y_flexible_test - flexible_model.predict(X_flexible_test)) ** 2).mean()
flexible_testing_R2 = 1 - flexible_testing_mse / ((y_flexible_test - y_flexible_test.mean()) ** 2).mean()

print(f"Mean square error of flexible model: {flexible_testing_mse: .4f}")
print(f"R2 of flexible model: {flexible_testing_R2: .4f}")

Mean square error of flexible model:  0.2647
R2 of flexible model:  0.2032


In [29]:
extra_flexible_training_mse = ((y_extra_flexible_train - extra_flexible_model.predict(X_extra_flexible_train)) ** 2).mean()
extra_flexible_training_R2 = 1 - extra_flexible_training_mse / ((y_extra_flexible_train - y_extra_flexible_train.mean()) ** 2).mean()
extra_flexible_training_adjR2 = 1 - extra_flexible_training_mse * y_extra_flexible_train.shape[0] / (y_extra_flexible_train.shape[0] - extra_flexible_model.n_features_in_) / ((y_extra_flexible_train - y_extra_flexible_train.mean()) ** 2).mean()

print(f"Mean square error of extra flexible model: {extra_flexible_training_mse: .4f}")
print(f"R2 of extra flexible model: {extra_flexible_training_R2: .4f}")
print(f"Adjusted R2 of extra flexible model: {extra_flexible_training_adjR2: .4f}")

Mean square error of extra flexible model:  0.1734
R2 of extra flexible model:  0.4736
Adjusted R2 of extra flexible model:  0.3117


In [30]:
extra_flexible_testing_mse = ((y_extra_flexible_test - extra_flexible_model.predict(X_extra_flexible_test)) ** 2).mean()
extra_flexible_testing_R2 = 1 - extra_flexible_testing_mse / ((y_extra_flexible_test - y_extra_flexible_test.mean()) ** 2).mean()

print(f"Mean square error of extra flexible model: {extra_flexible_testing_mse: .4f}")
print(f"R2 of extra flexible model: {extra_flexible_testing_R2: .4f}")

Mean square error of extra flexible model:  0.2975
R2 of extra flexible model:  0.0343
